src/components/select/CustomSelectDropdown.tsx

```diff
@@ -86,47 +89,69 @@ const CustomSelectDropdown: FC<Props> = ({
   useEffect(() => {
-    if (selectedIndex !== -1 && optionsRef.current[selectedIndex]) {
+    if (optionsRef.current[selectedIndex]) {
       optionsRef.current[selectedIndex].scrollIntoView({
         block: "nearest",
         inline: "nearest",
       });
     }
   }, [selectedIndex]);
 
+  const filteredOptions = options?.filter((option) => {
+    if (!search) return true;
+    const searchText = getOptionText(option) || option.value;
+    return searchText.toLowerCase().includes(search);
+  });
+
+  const getNextOptionIndex = (goingUp: boolean, prevIndex: number) => {
+    const increment = goingUp ? -1 : 1;
+    let currIndex = prevIndex + increment;
+    // skip disabled options for key board action
+    while (filteredOptions[currIndex] && filteredOptions[currIndex]?.disabled) {
+      currIndex += increment;
+    }
+
+    // consider upper bound for navigating down the list
+    if (increment > 0) {
+      return currIndex < filteredOptions.length ? currIndex : prevIndex;
+    }
+
+    // consider lower bound for navigating up the list
+    return currIndex >= 0 ? currIndex : prevIndex;
+  };
+
   // handle keyboard actions for navigating the select dropdown
   const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
     const upDownKeys = ["ArrowUp", "ArrowDown"];
 
     // prevent default browser actions for up, down, enter and escape keys
     // also prevent any other event listeners from being called up the DOM tree
-    if ([...upDownKeys, "Enter", "Escape"].includes(event.key)) {
+    if ([...upDownKeys, "Enter", "Escape", "Tab"].includes(event.key)) {
       event.preventDefault();
       event.nativeEvent.stopImmediatePropagation();
     }
 
     if (upDownKeys.includes(event.key)) {
       setSelectedIndex((prevIndex) => {
         const goingUp = event.key === "ArrowUp";
-        const increment = goingUp ? -1 : 1;
-        let currIndex = prevIndex + increment;
-        // skip disabled options for key board action
-        while (options[currIndex] && options[currIndex]?.disabled) {
-          currIndex += increment;
-        }
+        return getNextOptionIndex(goingUp, prevIndex);
+      });
+    }
 
-        // consider upper bound for navigating down the list
-        if (increment > 0) {
-          return currIndex < options.length ? currIndex : prevIndex;
+    if (event.key === "Tab") {
+      setSelectedIndex((prevIndex) => {
+        const goingUp = event.shiftKey;
+        if (prevIndex === 0 && goingUp) {
+          onClose();
+          return prevIndex;
         }
 
-        // consider lower bound for navigating up the list
-        return currIndex >= 0 ? currIndex : prevIndex;
+        return getNextOptionIndex(goingUp, prevIndex);
       });
     }
 
-    if (event.key === "Enter" && selectedIndex !== -1) {
-      onSelect(options[selectedIndex].value);
+    if (event.key === "Enter") {
+      onSelect(filteredOptions[selectedIndex].value);
```

Comment:
I think we have to double-check here if the selectedIndex is valid. Though for the permission selector this behaves correctly already.

Reply:
Since `selectedIndex` starts at 0 now, it will be a valid index. Note `filteredOptions` is computed before this part based on the search input state, and every time the search is updated we will reset the `selecteIndex` to 0.

Reply:
What if there are no search value matches and `filteredOptions` is empty?

Reply:
yes of course :+1: added a validation here now